sherpa/astro/utils/src/_utils.cc

```diff
@@ -338,25 +338,71 @@ namespace sherpa { namespace astro { namespace utils {
 			    &group ) )
       return NULL;
 
-    if( mask.get_size() == 0 ) {
-      PyErr_SetString( PyExc_TypeError,
-		       (char*)"mask array has no elements" );
+    if (mask.get_ndim() != 1) {
+      PyErr_SetString( PyExc_ValueError, (char*)"mask array must be 1D" );
+      return NULL;
+    }
+
+    if (group.get_ndim() != 1) {
+      PyErr_SetString( PyExc_ValueError, (char*)"group array must be 1D" );
+      return NULL;
+    }
+
+    // Use TypeError for the size=0 checks as this is what historically been
+    // used for this check (when only mask_size was checked).
+    //
+    npy_intp mask_size = mask.get_size();
+    npy_intp group_size = group.get_size();
+    if( mask_size == 0 ) {
+      PyErr_SetString( PyExc_TypeError, (char*)"mask array has no elements" );
+      return NULL;
+    }
+
+    if( group_size == 0 ) {
+      PyErr_SetString( PyExc_TypeError, (char*)"group array has no elements" );
+      return NULL;
+    }
+
+    if (group[0] < 0) {
+      PyErr_SetString( PyExc_ValueError, (char*)"The first element of group is negative" );
       return NULL;
     }
 
     if( EXIT_SUCCESS != res.zeros( group.get_ndim(), group.get_dims() ))
       return NULL;
 
-    npy_intp size = group.get_size();
     npy_intp jj = 0;
 
-    for( npy_intp ii = 0; ii < size; ++ii ) {
-      if( group[ ii ] >= 0 && ii != 0 )
+    // The group array starts a group with a value >= 0 and continues the
+    // group with a value < 0. The first element therefore starts a new
+    // group.
+    //
+    // The assumption is that the difference, in time, to just setting
+    //     res[ii] = mask[jj]
+    // is insignificant.
+    //
+    if( mask[jj] )
```

Comment:
While this is correct, it is safer to explicitly   if (mask[0])    here to aid in readability and to avoid accidentally introducing unintended behavior if code is added above this point which changes jj.

Reply:
Can you suggest what should be said or done here, as I'm not sure what you want? I have to admit that as I wrote this text last April I've forgotten exactly what I was thinking with these comments, so exactly what's meant to happen when the first mask value is < 0 (I claim that the data is then invalid, but we still may have to deal with it), or exactly what the comment about timing is really meant to say is now unclear.

Reply:
I'm going to assume this is just about the fact I'd stupidly used `jj` as the argument rather than `0` (and has now been addressed).